[✨feat] Scrap(Filter) 구현

jsoonworld opened the pull request:
# 📄 Work Description  
User와 Color를 기반으로 한 `Scrap` 엔티티를 정의했습니다.  
JPA 매핑을 포함한 도메인 설계를 적용했으며, 정적 팩터리 메서드(`of`)를 통해 의미 있는 방식으로 객체를 생성하도록 구성했습니다.  
색상 변경 기능(`changeColor`)과 HEX 문자열 반환 기능도 도메인 메시지로 노출했습니다.

---

# 💭 Thoughts  
Scrap이 생성되는 방식에 대해 고민하다가, 외부 객체(User 등)가 생성 책임을 갖기보단 Scrap 자신이 생성 책임을 갖는 구조가 더 자연스럽다고 판단했습니다.  
따라서 정적 팩터리 메서드를 통해 도메인 생성 책임을 Scrap 내부에 유지했습니다.  
또한 `equals/hashCode`는 ID를 기준으로 정의하여 JPA/Hibernate에서 안전하게 동작하도록 구성했습니다.  
별도로 검증할 도메인 로직이 없기 때문에 테스트는 작성하지 않았습니다.

---

# ✅ Testing Result  
- 현재 도메인 로직이 단순하며 상태 변경 외 검증 로직이 없어 테스트는 생략했습니다.

---

# 🗂 Related Issue  
- closed #24

jsoonworld committed:
feat(Scrap) : Scrap 엔티티 정의 및 생성 로직 구현

User와 Color를 기반으로 한 Scrap 엔티티를 정의하고,
정적 팩터리 메서드와 색상 변경 기능을 포함한 도메인 로직을 구현했습니다.

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/domain/scrap/Scrap.kt:
```diff
@@ -0,0 +1,53 @@
+        ): Scrap = Scrap(user = user, color = color)
+    }
+
+    fun changeColor(to: Color) {
```
넵! 이번 메서드들은 비교적 단순한 동작이라 문서 없이도 충분히 명확하다고 판단했지만,
향후 복잡한 로직이나 도메인 메서드가 추가될 경우 KDoc 주석도 함께 고민해보겠습니다. 🙂
좋은 제안 감사합니다!

jsoonworld committed:
ci(github): trigger GitHub Actions

jsoonworld committed:
Merge remote-tracking branch 'origin/develop' into feat/#24

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/domain/scrap/Scrap.kt:
```diff
@@ -0,0 +1,53 @@
+
+@Entity
+@Table(name = "scraps")
+class Scrap private constructor(
```
제안 감사합니다! 🙏  
말씀하신 내용처럼 JPA는 기본 생성자(no-arg constructor)가 필요하지만,  
현재 프로젝트에는 이미 `kotlin("plugin.jpa")` 플러그인이 적용되어 있어  
`@Entity` 클래스에 대해 `protected` no-arg 생성자가 자동으로 생성되고 있습니다.  

따라서 `Scrap` 클래스도 `private constructor`만으로 충분히 JPA가 사용할 수 있고,  
의도한 도메인 불변성도 잘 유지할 수 있어 별도 기본 생성자를 명시하지 않아도 괜찮습니다.  
좋은 피드백 감사합니다! 🙌

jsoonworld committed:
Merge remote-tracking branch 'origin/feat/#22' into feat/#24

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/scrap/Scrap.kt:
```diff
@@ -0,0 +1,53 @@
+@Entity
+@Table(name = "scraps")
+class Scrap private constructor(
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    val id: Long? = null,
```
처음에 이 부분을 보고 PK는 null이면 안 되는 거 아닌가 생각했었는데요..! 
찾아보니 `GenerationType.IDENTITY`로 기본키 생성을 데이터베이스에 위임하면서 엔티티 생성 시에는 null이어도 괜찮은 것 같더라구요 덕분에 알아갑니다~!

jsoonworld replied:
맞아요! 처음 보면 "PK가 null이어도 되나?" 싶은데,  
`GenerationType.IDENTITY` 전략에서는 DB가 ID를 자동으로 생성해주기 때문에  
엔티티를 생성할 때는 `id`가 null이어도 전혀 문제되지 않더라구요.

그리고 JPA에서는 `id`가 null이어야 해당 객체를 새 엔티티로 인식해서 `persist`를 수행해요.  
값이 이미 있으면 merge 대상으로 오해할 수도 있어서요!

말씀하신 것처럼 JPA는 내부적으로 리플렉션을 사용해서 객체를 만들기 때문에,  
`protected` 기본 생성자가 필요한 것도 중요한 포인트인 것 같아요 🙌  
저도 이 부분은 JPA 할 때마다 한 번씩 다시 확인하게 되는 부분이라 공감돼요😊

leeeyubin replied:
자세한 설명 감사합니다!!🚀

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/scrap/Scrap.kt:
```diff
@@ -0,0 +1,53 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    val id: Long? = null,
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", nullable = false)
+    val user: User,
```
제가 찾아봤을 때 `@JoinColumn`은 외래 키 칼럼을 설정해주는 속성이라고 알고 있는데요..!
저희의 ERD를 보면 Scraps 테이블에 Users 외래 키가 `userId`로 되어있어서요! 
스네이크 케이스랑 카멜 케이스가 호환이 되는 걸까요..?

jsoonworld replied:
앗, 말씀 듣고 ERD 다시 보니까 외래 키 컬럼명이 `user_id`가 아니라 `userId`로 되어 있었네요!  
DB 컬럼 이름과 정확히 맞춰주기 위해 `@JoinColumn(name = "userId")`로 수정하는 게 맞을 것 같아요 😊

스네이크 케이스로 착각하기 쉬운 부분인데, 짚어주셔서 덕분에 바로잡을 수 있었어요! 감사합니다 🙏

leeeyubin reviewed (approved):
엔티티도 뚝딱 만들어내는 장순님 멋집니다!! 🚀🚀

jsoonworld committed:
refactor(Scrap) : @JoinColumn 외래 키 컬럼명을 userId로 수정

- ERD 상 scraps 테이블의 외래 키 컬럼명이 user_id가 아닌 userId로 명시되어 있어,
  JPA 매핑 오류 방지를 위해 name 값을 "userId"로 수정

jsoonworld committed:
merge(feat/#22 → feat/#24) : Color 변경사항 반영
